dms/models/ir_attachment.py

```diff
@@ -54,21 +51,37 @@ def _dms_operations(self):
                 directories = self._get_dms_directories(
                     attachment.res_model, attachment.res_id
                 )
-            # Auto-create_files
+            # Auto-create_files (if not exists)
             for directory in directories:
-                self.env["dms.file"].sudo().create(
-                    {
-                        "name": attachment.name,
-                        "directory_id": directory.id,
-                        "attachment_id": attachment.id,
-                        "res_model": attachment.res_model,
-                        "res_id": attachment.res_id,
-                    }
+                dms_file_model = self.env["dms.file"].sudo()
+                dms_file = dms_file_model.search(
+                    [
+                        ("directory_id", "=", directory.id),
+                        ("res_model", "=", attachment.res_model),
+                        ("res_id", "=", attachment.res_id),
+                    ]
                 )
+                if not dms_file:
+                    dms_file_model.create(
+                        {
+                            "name": attachment.name,
+                            "directory_id": directory.id,
+                            "attachment_id": attachment.id,
+                            "res_model": attachment.res_model,
+                            "res_id": attachment.res_id,
+                        }
+                    )
 
     @api.model_create_multi
     def create(self, vals_list):
         records = super().create(vals_list)
         if not self.env.context.get("dms_file"):
             records._dms_operations()
         return records
+
+    @api.multi
+    def write(self, vals):
+        res = super().write(vals)
+        if not self.env.context.get("dms_file"):
+            self.with_context(attachment_action_write=True)._dms_operations()
```

Comment:
If this is done to differentiate attachments that come from a message and get attached to a record, shouldn't we check first than the `res_model,res_id` combination has changed, and trigger the DMS stuff only on that case?

If somebody goes and just changes the attachment name, would we really want to trigger this? How about some module that edits an attachment in some other form? Another alternative: is there a context key that tells us this write is being issued from the mail system?

Reply:
I have changed to use `mail_post_autofollow` context because exists in mail process and it's a better context.